jserv/semu
# Implement VirtIO GPU and input devices

## Summary

This pull request introduces initial support for `virtio-gpu` and `virtio-input` devices in `semu`, conforming to the [VirtIO specification 1.3](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html).

### VirtIO-GPU

* Based on `VirtIO MMIO transport layer` (i.e., Linux platform device)
* Implements basic 2D rendering support.
* Provides two rendering backends:
    * **Software-based** backend (supports 2D rendering only, ready to be integrated into [rv32emu](https://github.com/sysprog21/rv32emu)).
    * **virglrenderer** backend (to enable hardware-accelerated 3D rendering in a future patch).
* Compatible with:
    * [DirectFB2](https://github.com/directfb2/DirectFB2)
    * [X11 Window System](https://www.x.org/wiki/)
* Successfully tested with:
    * [kmscube](https://gitlab.freedesktop.org/mesa/kmscube)
    * [glmark2](https://github.com/glmark2/glmark2)
    * [glxgears](https://linuxreviews.org/Glxgears#google_vignette)

### VirtIO-Input

* Implements basic keyboard emulation.
* Implements basic mouse emulation.

## Prerequisites

> Note: this list is not yet exhaustive

```shell
$ sudo apt install libsdl2-dev libsdl2-2.0-0 libsdl2-image-dev libsdl2-image-2.0-0
```

## Build kernel image and test rootfs image

```shell
$ make rootfs.cpio # Download prebuilt minimal rootfs for initrd
$ ./scripts/build-image.sh --all --x11 --external-root --extra-packages
```

Use `--x11` for full setup; omit it for a minimal setup.

## Run `X11`

1. Launch `semu`:

```shell
$ make check
```

2. Run `startx`

```
$ startx
```

3. After `X11` shows up, launch the benchmark programs:

```
$ glmark2
$ glxgears
```

## Run `DirectFB2`

1. Launch `semu`:

```shell
$ make check
```

2. Kill `X11` (Skip if `--x11` is not used)

```
$ ps | grep 'Xorg'
   <PID> root     /usr/bin/Xorg :0.0 vt01 -s 0 -noreset -allowMouseOpenFail
$ kill <PID>
```

3. Set up environment variables via `run.sh`:

```shell
$ source ./run.sh
```

4. Run `DirectFB2` test programs:

```
df_andi        df_fire df_neo     df_window
df_texture     df_video           df_knuckles
df_input       df_neo             df_window
df_matrix      df_dok             df_fire
df_palette     df_cpuload         df_spacedream
df_layers      df_pss             df_input
```

## Run `kmscube`

1. Launch `semu`:

```shell
$ make check
```

2. Kill `X11` (Skip if `--x11` is not used)

```
$ ps | grep 'Xorg'
   <PID> root     /usr/bin/Xorg :0.0 vt01 -s 0 -noreset -allowMouseOpenFail
$ kill <PID>
```

3. Run `kmscube`

```
$ kmscube
```

## Run `modetest`

The simplest method to test VirtIO-GPU.

1. Enable `BR2_PACKAGE_LIBDRM_INSTALL_TESTS` in the Buildroot then rebuild the image.

2. Launch `semu`:

```shell
$ make check
```

3. Run `modetest` program

```shell
$ modetest -M virtio_gpu -s 34:1024x768
```

A test image should then show up in the display window.
